tests/stdlib/ttext_layouter.nim

```diff
@@ -0,0 +1,205 @@
+       argument5, argument6
+       argument7, argument8
+       argument9, argument10)
+AVeryLongAndDescriptiveFunctionName(argument1
+                                    argument2
+                                    argument3
+                                    argument4
+                                    argument5
+                                    argument6
+                                    argument7
+                                    argument8
+                                    argument9
+                                    argument10)
+AVeryLongAndDescriptiveFunctionName(
+    argument1, argument2, argument3, argument4
+    argument5, argument6, argument7, argument8
+    argument9, argument10
+)
+stmtPragmas* = { wChecks
+                 wOverflowChecks
+                 wNilChecks }
+stmtPragmas* = {
+  wChecks,      wObjChecks,
+  wBoundChecks, wOverflowChecks, wNilChecks
+}
+block:
+  var i: int = 0
+    postInc(i)
+'''
+"""
+
+import experimental/text_layouter
+import std/[strutils, sequtils, strformat]
+
+initBlockFormatDSL()
+
+type
+  StrStore = ref object
+    strings: seq[string]
+
+proc str(store: var StrStore, str: string): LytStr =
+  result = lytStrIdx(store.strings.len, str.len)
+  store.strings.add str
+
+proc str(s: StrStore, str: LytStr): string =
+  if str.isSpaces():
+    repeat(" ", str.len)
+
+  else:
+    s.strings[str.id.toIndex()]
+
+proc toString(s: StrStore, opts: LytOptions, blc: LytBlock): string =
+  let lyt = blc.toLayout(opts)
+  for event in formatEvents(lyt):
+    case event.kind:
+      of layEvNewline:
+        result.add "\n"
+
+      of layEvSpaces:
+        result.add repeat(" ", event.spaces)
+
+      of layEvStr:
+        result.add s.str(event.str)
+
+  var res: seq[string]
+  for line in splitLines(result):
+    res.add strip(line, leading = false)
+
+  return res.join("\n")
+
+var s = StrStore()
+
+
+proc toString(b: LytBlock, width: int = 80): string =
+  var opts = initLytOptions()
+  opts.rightMargin = width
+  toString(s, opts, b)
+
+proc toLyt(b: LytBlock): Layout = b.toLayout(initLytOptions())
+
+proc lTX(str: string): LytBlock = lT(s.str(str))
+
+proc lytProc(args: openArray[LytBlock], body: LytBlock): LytBlock =
+  let
+    h = "proc ("
+    t = ") = "
+    hsep = initHSeparated(@args, lTX(", "))
+    vsep = initVSeparated(@args, lTX(", "))
+
+
+  result = lC(
+    lH(lTX(h), hsep, lTX(t), body),
+    lV(lH(lTX(h), hsep, lTX(t), lI(2, body))),
+    lV(lTX(h), lI(4, vsep), lTX(t), lI(2, body))
+  )
+
+
+
+for args in [1, 2]:
+  for body in [20, 60]:
+    let bl = lytProc(
+      args = mapIt(0 ..< args, lTX(&"arg{it}: arg{it}_type")),
+      body = lTX(repeat("?", body))
+    )
+
+    echo toString(bl)
+
+if true:
+  for i in [1, 5, 10]:
+    var blocks = mapIt(0 .. i, lTX(&"arg{it}: int{it}"))
+    let bl = lH(lTX("proc ("),
+      lC(join(lH(blocks), lTX(", ")), join(lV(blocks), lTX(","))),
+      lTX(")"))
+
+    echo toString(bl)
+
+let comm = s.str(", ")
+
+echo toString(
```

Comment:
Aside from the ones inside `for`, are these `echo` ammenable to being transformed into `doAssert`?

Reply:
oh, not easily. Just noticed the toString definition above.

Reply:
Yes, will do now

Reply:
moved all checks into doAssert